src/vm/tests/events.rs

```diff
@@ -0,0 +1,310 @@
+    };
+}
+
+#[test]
+fn test_emit_print_nok() {
+    let contract =
+        "(define-public (emit-event-nok)
+            (begin
+                (print \"Hello world\")
+                (err u1)))";
+
+    let (value, events) = helper_execute(contract, "emit-event-nok");
+    assert_eq!(value, Value::error(Value::UInt(1)).unwrap());
+    assert_eq!(events.len(), 0);
+}
+
+#[test]
+fn test_emit_stx_transfer_ok() {
+    let contract =
+        "(define-constant sender 'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR)
+        (define-constant recipient 'SM2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQVX8X0G)
+        (define-fungible-token token)
+        (define-public (emit-event-ok)
+            (begin
+                (unwrap-panic (stx-transfer? u10 sender recipient))
+                (ok u1)))";                
+                
+    let (value, mut events) = helper_execute(contract, "emit-event-ok");
+    assert_eq!(value, Value::okay(Value::UInt(1)).unwrap());
+    assert_eq!(events.len(), 1);
+    match events.pop() {
+        Some(StacksTransactionEvent::STXEvent(STXEventType::STXTransferEvent(data))) => {
+            assert_eq!(data.amount, 10u128);
+            assert_eq!(Value::Principal(data.sender), execute("'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR"));
+            assert_eq!(Value::Principal(data.recipient), execute("'SM2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQVX8X0G"));
+        },
+        _ => panic!("assertion failed")
+    };            
+}
+
+#[test]
+fn test_emit_stx_transfer_nok() {
+    let contract =
+        "(define-constant sender 'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR)
+        (define-constant recipient 'SM2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQVX8X0G)
+        (define-fungible-token token)
+        (define-public (emit-event-nok)
+            (begin
+                (unwrap-panic (stx-transfer? u10 sender recipient))
+                (err u1)))";                
+
+    let (value, events) = helper_execute(contract, "emit-event-nok");
+    assert_eq!(value, Value::error(Value::UInt(1)).unwrap());
+    assert_eq!(events.len(), 0);
+}
+
+#[test]
+fn test_emit_stx_burn_ok() {
+    let contract =
+        "(define-constant sender 'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR)
+        (define-fungible-token token)
+        (define-public (emit-event-ok)
+            (begin
+                (unwrap-panic (stx-burn? u10 sender))
+                (ok u1)))";                
+                
+    let (value, mut events) = helper_execute(contract, "emit-event-ok");
+    assert_eq!(value, Value::okay(Value::UInt(1)).unwrap());
+    assert_eq!(events.len(), 1);
+    match events.pop() {
+        Some(StacksTransactionEvent::STXEvent(STXEventType::STXBurnEvent(data))) => {
+            assert_eq!(data.amount, 10u128);
+            assert_eq!(Value::Principal(data.sender), execute("'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR"));
+        },
+        _ => panic!("assertion failed")
+    };
+}
+
+#[test]
+fn test_emit_stx_burn_nok() {
```

Comment:
Can you add a test that has a nested public function that aborts? For example:

```
(define-public (emit-event-nok)
   (begin
      (print "bar")
      (err u1)))
(define-public (emit-event-ok)
    (begin
       (emit-event-nok)
       (print "foo")
       (ok 1)))
```

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1319/commits/911d209a9e885f5bf5fda164e8c1a2f62017ba84